src/libraries/System.Private.CoreLib/src/System/SpanSplitEnumerator.T.cs

```diff
@@ -0,0 +1,68 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    /// <summary>
+    /// <see cref="System.SpanSplitEnumerator{T}"/> allows for enumeration of each element within a <see cref="System.ReadOnlySpan{T}"/>
+    /// that has been split using a provided separator.
+    /// </summary>
+    public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
+    {
+        private readonly ReadOnlySpan<T> _buffer;
+        private readonly ReadOnlySpan<T> _separatorSequence;
+        private readonly T _separator;
+        private readonly bool _isSequence;
+        private readonly int _separatorLength;
+        private int _offset;
+        private int _index;
+
+        /// <summary>
+        /// Returns an enumerator that allows for iteration over the split span.
+        /// </summary>
+        /// <returns>Returns a <see cref="System.SpanSplitEnumerator{T}"/> that can be used to iterate over the split span.</returns>
+        public SpanSplitEnumerator<T> GetEnumerator() => this;
+
+        /// <summary>
+        /// Returns the current element of the enumeration.
+        /// </summary>
+        /// <returns>Returns a <see cref="System.Range"/> struct that defines the bounds of the current element withing the source span.</returns>
+        public readonly Range Current => new Range(_offset, _offset + _index - _separatorLength);
```

Comment:
This calculation is not easy to follow and fails for some edge cases (highlighted in my other comment).

Here's an implementation I made atop your work that I think is easier to grok and handles edge-cases. In particular, prefer for the `Current` property to return a simpler expression:

````cs
public Range Current => new Range(_start, _end);
````

```cs
public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
{
    private readonly ReadOnlySpan<T> _buffer;

    private readonly bool _isSingleToken;

    // Only one of these used.
    private readonly ReadOnlySpan<T> _separatorSequence;
    private readonly T _separator;

    private readonly int _separatorLength;
    private int _offset;

    private int _start;
    private int _end;

    private readonly bool _initialized;

    public SpanSplitEnumerator<T> GetEnumerator() => this;

    internal SpanSplitEnumerator(ReadOnlySpan<T> span, T separator)
    {
        _buffer = span;

        _isSingleToken = true;
        _separator = separator;
        _separatorSequence = default;
        _separatorLength = 1;

        _start = 0;
        _end = 0;
        _offset = 0;

        _initialized = true;
    }

    internal SpanSplitEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> separator)
    {
        _buffer = span;

        _isSingleToken = false;
        _separatorSequence = separator;
        _separator = default!;
        _separatorLength = separator.Length;

        _start = 0;
        _end = 0;
        _offset = 0;

        _initialized = true;
    }

    public Range Current => new Range(_start, _end);

    public bool MoveNext()
    {
        if (!_initialized || _offset > _buffer.Length)
        {
            return false;
        }

        ReadOnlySpan<T> slice = _buffer.Slice(_offset);
        _start = _offset;

        int separatorIndex = _isSingleToken ? slice.IndexOf(_separator) : slice.IndexOf(_separatorSequence);
        int subStringLength = separatorIndex == -1 ? _buffer.Length - _start : separatorIndex;

        _end = _start + subStringLength;
        _offset = _end + (_separatorLength == 0 ? 1 : _separatorLength);

        return true;
    }
}
```